interfaces/builtin/custom_device_test.go

```diff
@@ -0,0 +1,364 @@
+			`snap "provider" has interface "custom-device" with invalid value type string for "udev-tagging" attribute.*`,
+		},
+		{
+			"udev-tagging:\n    - foo: bar}",
+			`custom-device "udev-tagging" invalid "foo" tag: unknown key`,
+		},
+		{
+			"udev-tagging:\n    - subsystem: 12",
+			`custom-device "udev-tagging" invalid "subsystem" tag: value "12" is not a string`,
+		},
+		{
+			"udev-tagging:\n    - subsystem: deal{which,this}",
+			`custom-device "udev-tagging" invalid "subsystem" tag: value "deal{which,this}" contains invalid characters`,
+		},
+		{
+			"udev-tagging:\n    - subsystem: bar",
+			`custom-device udev tagging rule missing mandatory "kernel" key`,
+		},
+		{
+			"udev-tagging:\n    - kernel: bar",
+			`custom-device "udev-tagging" invalid "kernel" tag: "bar" does not match a specified device`,
+		},
+		{
+			"udev-tagging:\n    - attributes: foo",
+			`custom-device "udev-tagging" invalid "attributes" tag: value "foo" is not a map`,
+		},
+		{
+			"udev-tagging:\n    - attributes: {key\": noquotes}",
+			`custom-device "udev-tagging" invalid "attributes" tag: key "key"" contains invalid characters`,
+		},
+		{
+			"udev-tagging:\n    - environment: {key: \"va{ue}\"}",
+			`custom-device "udev-tagging" invalid "environment" tag: value "va{ue}" contains invalid characters`,
+		},
+	}
+
+	for _, testData := range data {
+		snapYaml := fmt.Sprintf(customDeviceYaml, testData.slotYaml)
+		_, slot := MockConnectedSlot(c, snapYaml, nil, "hwdev")
+		err := interfaces.BeforePrepareSlot(s.iface, slot)
+		c.Check(err, ErrorMatches, testData.expectedError, Commentf("yaml: %s", testData.slotYaml))
+	}
+}
+
+func (s *CustomDeviceInterfaceSuite) TestSlotAttribute(c *C) {
+	snapYaml := `name: consumer
+version: 0
+slots:
+ hwdev:
+  interface: custom-device
+`
+	_, slot := MockConnectedSlot(c, snapYaml, nil, "hwdev")
+	err := interfaces.BeforePrepareSlot(s.iface, slot)
+	c.Assert(err, IsNil)
+	c.Check(slot.Attrs["custom-device"], Equals, "hwdev")
+}
+
+func (s *CustomDeviceInterfaceSuite) TestStaticInfo(c *C) {
+	si := interfaces.StaticInfoOf(s.iface)
+	c.Check(si.ImplicitOnCore, Equals, false)
+	c.Check(si.ImplicitOnClassic, Equals, false)
+	c.Check(si.Summary, Equals, `provides access to custom devices specified via the gadget snap`)
+	c.Check(si.BaseDeclarationSlots, testutil.Contains, "custom-device")
+}
+
+func (s *CustomDeviceInterfaceSuite) TestAppArmorSpec(c *C) {
+	spec := &apparmor.Specification{}
+
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	plugSnippet := spec.SnippetForTag("snap.consumer.app")
+
+	c.Assert(spec.AddConnectedSlot(s.iface, s.plug, s.slot), IsNil)
+	slotSnippet := spec.SnippetForTag("snap.provider.app")
+
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+
+	c.Check(plugSnippet, testutil.Contains, `"/dev/input/event[0-9]" rw,`)
+	c.Check(plugSnippet, testutil.Contains, `"/dev/input/mice" rw,`)
+	c.Check(plugSnippet, testutil.Contains, `"/dev/js*" r,`)
```

Comment:
I thought that `*` was going to get mapped in the AppArmor allowed filepaths to `**` ?

Reply:
Initially yes, but Samuele preferred to keep the mapping to `*`.